### ISSUE TITLE
Use in node

### ISSUE THREAD

**Issue by lamplightdev** [matthewp/dom-shim#6]
Hey, just wondering if usage in node would be a desired goal of the this library?

The use of url parameters in determining the `global` and `props` parameters make using this library in contexts such as node (and other runtimes, e.g. CloudFlare Workers) problematic. I've had some level of success by bundling with rollup and https://www.npmjs.com/package/rollup-plugin-url-resolve, but there are some issues with `import.meta.url` not being rewritten properly.

Refactoring to export a function instead e.g. `export default function(props=[], global=false) {...}` would enable use in these contexts, at the expense of consumers of the library having to wrap the function in a further module so that the globals get set before other imports. Is that a trade-off you think is worth making?

Cheers

**Comment by matthewp** [comment]
Definitely! I don't want to change the core API, but since modules are free we could have a separate entrypoint that is as you described (maybe shim.js?), then have mod.js call that function.

**Comment by lamplightdev** [comment]
Great, I will have a go at this then.

**Comment by matthewp** [comment]
Is the fact that has `import.meta.url` in the code the problem or is the actual dependency on that the problem? Because you can already omit `global` and do this:

```js
import { shim } from 'https://cdn.spooky.click/dom-shim/1.0.0/mod.js';

shim();
```

So if it's just a matter of setting `props` we could maybe have a second export, something like `setupValues` that creates the `shimValues` property, so you could wind up doing this:

```js
import { shim, setupValues } from 'https://cdn.spooky.click/dom-shim/1.0.0/mod.js';

setupValues(['HTMLElement', 'document']);
shim();
```

And avoid using the query params, if that's the goal here. That avoids the extra module.

**Comment by lamplightdev** [comment]
The problem is the `new URL(import.meta.url)` will choke if the URL's not valid, so I'm not sure that would work?

Here's my first pass at a solution: https://github.com/lamplightdev/dom-shim/tree/node

It maintains the current behaviour in Deno with `mod.js`, and in node `shim.js` would need to be imported directly and the params passed into the returned function.

I've also had a go at how Ocean might work in node with these changes: https://github.com/lamplightdev/ocean/tree/node-shim

Again the current behaviour is maintained, but in node you'd have to `import 'http://.../ocean/lib/shim-node.js'` which will add to the global scope, and the user would have to `unshim` manually if required using, `import { unshim } from 'http://.../ocean/lib/shim-node.js'`

Not sure it's the most elegant solution yet, but it does work :)

**Comment by matthewp** [comment]
Looking good! I'll have better feedback probably tomorrow. The only thing I see (from the Ocean branch) that's an issue is that I think we should keep the name "node" out of filenames as both Ocean and dom-shim are meant to be universal JS. No suggestions yet though, I'll take a closer look at what the code is doing tomorrow.

**Comment by matthewp** [comment]
Ok, hopefully some better feedback:

# dom-shim

I'm pretty much on board with how this works, a couple of minorish points:

1. There's some patching that we do like adding `requestAnimationFrame` that should probably be in another module just to keep things organized, but this is unrelated to your change so I can do that at another time.
2. Since the setup functions returns a `shim` method, do we really need/want the `global` argument? The caller can just call shim() itself, I think that's a little cleaner and gives this module more of a purpose (setup a new DOM environment).
3. Now I'm kind of thinking this should be a class! I'm happy to do this one myself as a follow-up though.

# ocean

I think I follow this for the most part, maybe some naming and/or organization would help, but can tackle this after the dom-shim changes are in.

**Comment by lamplightdev** [comment]
Thanks for the pointers! I'll take a look in the week.

**Comment by lamplightdev** [comment]
Do you mean remove the `global` parameter altogether? I see it's not really needed in the node context as you can export `shim` and call that, but in the deno context you could no longer have the direct import. i.e.

```js
import 'https://cdn.spooky.click/ocean/1.2.6/shim.js?global';
import { Ocean } from 'https://cdn.spooky.click/ocean/1.2.6/mod.js';
```

would have to be replaced with some kind of intermediate import in order to apply the shim before the other imports are processed:

```js
// setup.js
import { shim, unshim } from 'https://cdn.spooky.click/ocean/1.2.6/shim.js';
shim();

export { unshim };
```

and

```js
import { unshim } from './setup.js';
import { Ocean } from 'https://cdn.spooky.click/ocean/1.2.6/mod.js';
```

Or is that acceptable?

**Comment by matthewp** [comment]
@lamplightdev No, It would still nice to have the ?global parameter for environments that can import URLs. mean remove it as an argument from the setup function that you were creating: https://github.com/lamplightdev/dom-shim/blob/1aedf5307ca5989270327474be7f85ecad4349e8/lib/shim.js#L12

Since the user already will have access to a `shim()` function that is returned from that, they can just call that if they want to shim the global.

**Comment by lamplightdev** [comment]
I think that function still needs a `global` parameter as it's used in `mod.js`?

**Comment by matthewp** [comment]
@lamplightdev I don't think so, mod.js can just call shim() can't it?

**Comment by matthewp** [comment]
Happy to work this out via a PR if that would be easier, I can submit suggestions.

**Comment by matthewp** [comment]
Closed by #7 and #8
